### ISSUE TITLE
Widget Visibility: existing "Post Type" rules are removed in 4.7

### ISSUE THREAD

**Issue by jeherve** [Automattic/jetpack#6595]
#### Steps to reproduce the issue

1. Start from a site running Jetpack 4.6.
2. Go to Appearance > Widgets and set up some widgets with custom visibility rules.
3. Use the "Post Type" option for one of the rules.

![screen shot 2017-03-07 at 22 27 21](https://cloud.githubusercontent.com/assets/426388/23678994/91d41102-0386-11e7-90bb-c727b2e95a47.png)

4. Update to Jetpack 4.7.
5. That option can't be displayed anymore:

![screen shot 2017-03-07 at 22 35 42](https://cloud.githubusercontent.com/assets/426388/23678990/8d993e0a-0386-11e7-8a7a-34c24cd2f22e.png)

Related: #6333

**Until the problem is fixed, one can work around the problem by choosing “Page”, then the Post type in the widget visibility rules.**

**Comment by jeherve** [comment]
Also reported in 3096698-t

**Comment by zinigor** [comment]
The rules were moved to the "Page" major rule, so I'd say that this is expected behaviour. The widgets still follow the old rules, and to fix their appearance here the user needs to change the major rule to "Page" -> "Post types" -> desired post type. Please open this issue again if there are additional troubles.

**Comment by jeherve** [comment]
> the user needs to change the major rule to "Page" -> "Post types" -> desired post

Is that something we could do on update for the user?

**Comment by zinigor** [comment]
> Is that something we could do on update for the user?

Good idea, I'll take a look. Let's reopen this.

**Comment by davidpagepsycle** [comment]
This has caused a problem for us. Whereas the rule for a singular post type was moved to the "Page" major rule, the rule for an archive of a post type has been lost.

Steps to reproduce:
1. Start from a site running Jetpack 4.6.
2. Go to Appearance > Widgets and set up some widgets with custom visibility rules.
3. Use the "Post Type" option for one of the rules, selecting the "Archive of [post-type]" as the minor rule.  
![screen shot 2017-03-09 at 12 42 39 pm](https://cloud.githubusercontent.com/assets/8223525/23750674/ef8e617c-04c5-11e7-8e80-16b9736e5b25.png)
4. Update to Jetpack 4.7.
5. Not only is the option now shown as blank, but it's not possible to recreate it as only singular post types are supported under "Page".
![screen shot 2017-03-09 at 12 34 04 pm](https://cloud.githubusercontent.com/assets/8223525/23750596/9b8ef4a6-04c5-11e7-86f3-c65ec8e91957.png)
6. Underneath the code still supports the old rule (see https://github.com/Automattic/jetpack/blob/master/modules/widget-visibility/widget-conditions.php#L584-L585)

**Comment by davidpagepsycle** [comment]
I've added a PR to restore the ability to set a condition for a custom Post Type Archive, though it doesn't change the issue of the old rule format for them to the new format.

**Comment by jeherve** [comment]
@davidpagepsycle Thanks, we'll take a look at #6614.

> it doesn't change the issue of the old rule format for them to the new format.

#6615 should take care of that.

**Comment by davidpagepsycle** [comment]
@jeherve I've just commented on #6615 as the migration of Post Type Archive rules will break those existing conditions.

**Comment by zinigor** [comment]
@davidpagepsycle do you mind if I cherry pick your commits into this PR?

**Comment by davidpagepsycle** [comment]
@zinigor Sure, just looking to restore the lost archive functionality. Have a client site that relies upon it.

**Comment by jeherve** [comment]
Reopening as the migration routine still needs some work.

**Comment by jeherve** [comment]
Reopening as I found some issues when testing #6633.

1. I merged the PR
2. I created a new build of `master-stable`
3. I created a brand new site where I installed Jetpack **4.6**.
4. I created a bunch of rules there.
5. I updated to Jetpack 4.7, and confirmed the bug reported above.
6. I then installed the Beta plugin and updated to the latest `master-stable` ( I confirmed that the changes introduced in #6633 were there. 
7 There are still missing rules in some cases.

This worked:

![screen shot 2017-03-13 at 17 48 12](https://cloud.githubusercontent.com/assets/426388/23866459/308db1d6-0819-11e7-807f-64223bc74f90.png)
![screen shot 2017-03-13 at 17 48 01](https://cloud.githubusercontent.com/assets/426388/23866462/34389580-0819-11e7-98bb-1169459a1bc6.png)
![screen shot 2017-03-13 at 17 47 32](https://cloud.githubusercontent.com/assets/426388/23866467/387d1dfa-0819-11e7-96f0-c8892515fed6.png)

This didn't:

![screen shot 2017-03-13 at 17 47 21](https://cloud.githubusercontent.com/assets/426388/23866474/3fc145a0-0819-11e7-8774-1dbd2ae671c4.png)
![screen shot 2017-03-13 at 17 47 44](https://cloud.githubusercontent.com/assets/426388/23866482/477adea0-0819-11e7-967f-0d2cf1b26107.png)

The rules I used are explained in each widget title.

**Comment by jeherve** [comment]
I ran a second test, and the migration worked well!

- Before: http://i.wpne.ws/jWmR
- After: http://i.wpne.ws/jXB8

Not sure what went wrong the first time. It would be nice to have someone else test just to be sure.

**Comment by eliorivero** [comment]
I tested it a few times and it's not working for me, the rules aren't migrated. Maybe I'm doing something wrong? These are the steps I followed:
1. `wp plugin install jetpack --version=4.6 --force --activate`
2. went to Widgets and setup WV rules, including Post Type rules
3. installed 4.7 by overwriting the folder manually in Finder
4. set Jetpack Beta **Use Jetpack Version:** to Bleeding Edge and updated
5. went to Widgets, checked WV rules, they were blank

**Comment by zinigor** [comment]
@eliorivero I think that the problem is in the step number 3, you need to go straight from 4.6 to bleeding edge. If you update to `branch-4.7`, there are no migration scripts, and when you upgrade to bleeding edge after that, there is no difference in versions, and the migration doesn't run.
It will not be the case for regular updates because users will update from 4.7 to 4.7.1.

**Comment by zinigor** [comment]
So it looks like the problem is finally solved. If anyone finds a problem with the rules, or the migration script, please reopen.
